### PR TITLE
Weight coefficent for remove negative spectrum value

### LINKED ISSUE #136: Negative NDBC directional energy spectrum
According to NOAA doc, directional spectrum could be negative due to trigonometric function. This could be an issue (for example, take log value). Here is the description from the NOAA. 
image
Do you think about add the optional feature to adjust those negative spectrum? The solution is fairly straightforward, introduced by this paper: _Use of advanced directional wave spectra analysis methods_
image

### PR DESCRIPTION
**Pull request by lubyant**
Close #136

### CODE REVIEW

**Comment by lubyant** [comment]
Description for weight the coefficient:

**Comment by lubyant** [comment]
@rsoutelino Please review this pr if you are available

**Comment by lubyant** [comment]
@rsoutelino would you please review this pr?

**Comment by rsoutelino** [comment]
@lubyant - you probably meant to request review from @rafa-guedes ?

File: wavespectra/input/ndbc.py [hunk 1]
```diff
@@ -48,22 +50,29 @@ def read_ndbc(url, directional=True, dd=10.0, chunks={}):
 
     """
     dset = xr.open_dataset(url).chunk(chunks)
-    return from_ndbc(dset, directional=directional, dd=dd)
+    return from_ndbc(dset, directional=directional, dd=dd, weight_coeff=weight_coeff)
 
 
-def _construct_spectra(ef, swd1, swd2, swr1, swr2, dir):
+def _construct_spectra(ef, swd1, swd2, swr1, swr2, dir, weight_coeff: bool = False):
```
Comment (daniel-caichac-DHI) [inline wavespectra/input/ndbc.py:56]: Question (not that I am reviewing or anything. but if the integral parameters remain, and this boolean solves the negative values, wouldn't it be better if this `weight_coeff` Defaults to True?

Comment (daniel-caichac-DHI) [reply wavespectra/input/ndbc.py:56]: Well, I guess @rafa-guedes needs to decide, but, just hear me out, thinking out loud here.

I've been thinking and I cannot think about any scenario where people (anyone) would actually _want_ to have negative energy density values. I imagine that the vast majority of people would appreciate that the quickes/simplest reconstruction of NDBC 2D Spectra is correct from a physical point (ie, all energy >= 0).  Which is why I thought about defaulting to True.

But maybe I am missing the big picture or some obvious case where someone would want the raw reconstruction ( IMO the negative values are just a mathematical artifact, hence not desired)

**Comment by rafa-guedes** [comment]
@lubyant thanks for this pull request! I'll have a closer look and merge it or get back to you in case there is any comment or question. I agree with @daniel-caichac-DHI that this option could be True by default - let's perhaps see what others here might think as well.

**Comment by rafa-guedes** [comment]
@lubyant the directional spreading seems to be influenced by the weighting coeff - is this expected? Below are a some comparisons of wave stats and spectra from the ndbc netcdf in the test files:

weight_coeff_efth

weight_coeff_timeseries

**Comment by rafa-guedes** [comment]
This is a sample code to reproduce the plots above:

```
    import matplotlib.pyplot as plt
    from wavespectra import read_ndbc

    ncfile = "/source/wavespectra/tests/sample_files/ndbc/42098w9999.nc"
    dset1 = read_ndbc(url=ncfile, directional=True, dd=5, weight_coeff=False)
    dset2 = read_ndbc(url=ncfile, directional=True, dd=5, weight_coeff=True)

    # Plot spectra
    t = "20150612T00"
    fig, axs = plt.subplots(1, 2, figsize=(15, 6))

    ds = dset1.sel(time=[t]).squeeze(drop=True).efth
    vmin, vmax = float(ds.min()), float(ds.max())
    ds.plot(ax=axs[0], cmap="turbo", vmin=vmin, vmax=vmax)
    axs[0].set_title("no weight_coeff")
    axs[0].set_ylabel("Frequency (Hz)")

    ds = dset2.sel(time=[t]).squeeze(drop=True).efth
    vmin, vmax = float(ds.min()), float(ds.max())
    ds.plot(ax=axs[1], cmap="turbo", vmin=vmin, vmax=vmax)
    plt.title("weight_coeff")
    axs[1].set_title("weight_coeff")
    axs[1].set_ylabel("")
    plt.savefig("weight_coeff_efth.png")

    # Plot timeseries
    ds1 = dset1.spec.stats(["hs", "tp", "dpm", "dspr"]).squeeze(drop=True)
    ds2 = dset2.spec.stats(["hs", "tp", "dpm", "dspr"]).squeeze(drop=True)
    fig, axs = plt.subplots(4, 1, figsize=(15, 10))
    ds1["hs"].plot(ax=axs[0], label="no weight_coeff", linewidth=3)
    ds2["hs"].plot(ax=axs[0], label="weight_coeff", linewidth=1)
    axs[0].set_ylabel("Hs (m)")
    ds1["tp"].plot(ax=axs[1], label="no weight_coeff", linewidth=3)
    ds2["tp"].plot(ax=axs[1], label="weight_coeff", linewidth=1)
    axs[1].set_ylabel("Tp (s)")
    ds1["dpm"].plot(ax=axs[2], label="no weight_coeff", linewidth=3)
    ds2["dpm"].plot(ax=axs[2], label="weight_coeff", linewidth=1)
    axs[2].set_ylabel("Dpm (degree)")
    ds1["dspr"].plot(ax=axs[3], label="no weight_coeff", linewidth=3)
    ds2["dspr"].plot(ax=axs[3], label="weight_coeff", linewidth=1)
    axs[3].set_ylabel("dspr (degree)")
    axs[3].legend()
    plt.savefig("weight_coeff_timeseries.png")
    plt.show()

```

**Comment by daniel-caichac-DHI** [comment]
I think it does, as per NDBC website
`These other parameters are ALPHA1 (mean wave direction), ALPHA2 (principle wave direction), and R1 and R2 (parameters which describe the directional spreading about the main direction).`
https://www.ndbc.noaa.gov/faq/wave.shtml

so if we are multiplying R1 and R2 by anything, the directional spreading changes.
I was trying to look at the timeseries of wave parameters of the buoy you plotted (https://www.ndbc.noaa.gov/view_text_file.php?filename=42098h2015.txt.gz&dir=data/historical/stdmet/) but unfortunately they do not provide Spreading.

An average spreading of ~60deg does sound a little bit large though, but the 30deg is calculated from "negative" energy values, so not sure how accurate that is.

**Comment by lubyant** [comment]
Yes, I think coefficient weighting could not deal with the directional spreading.

**Comment by rafa-guedes** [comment]
@lubyant @daniel-caichac-DHI this is finally merged, sorry for taking a bit long. I ended up leaving the new `weight_coeff` as False by default, just because of that caveat with the directional spreading. I'll be releasing a new version shortly including these changes. Thanks for your nice work!

**Pull request merged**